[redo] Use FunctionSchema instead of char* for dispatch

bwasti opened the pull request:
Previous PR was landed in a broken state

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28295 [redo] Use FunctionSchema instead of char* for dispatch**

Differential Revision: [D18066217](https://our.internmc.facebook.com/intern/diff/D18066217)

Commit:
Use FunctionSchema instead of char* for dispatch

yf225 commented:
@bwasti This PR seems to breaks `win-ws2016-cuda9-cudnn7-py3-build` on master with the following error:
```
Error: Internal Compiler Error (codegen): "there was an error in verifying the lgenfe output!"
```
I will have to revert it. The solution to this build error is at https://github.com/pytorch/pytorch/issues/12117 and https://github.com/pytorch/pytorch/pull/26246/files.

Commit:
Update on "[redo] Use FunctionSchema instead of char* for dispatch"

Previous PR was landed in a broken state


Differential Revision: [D18012804](https://our.internmc.facebook.com/intern/diff/D18012804)
